evm/src/ampleforth/AmpleforthSwapAdapter.sol

```diff
@@ -0,0 +1,435 @@
+                    IWAMPL(WAMPL).wrapperToUnderlying(specifiedAmounts[i]);
+                prices[i] = Fraction(outAmt, specifiedAmounts[i]);
+            }
+        } else {
+            revert Unavailable("SwapRouteUndefined");
+        }
+    }
+
+    /**
+     * @dev Computes price quotes for STAMPL swaps.
+     */
+    function _calculatePriceSTAMPL(
+        address sellToken,
+        address buyToken,
+        uint256[] memory specifiedAmounts,
+        Fraction[] memory prices
+    ) private {
+        if (sellToken == AMPL && buyToken == SPOT) {
+            // AMPL -> SPOT
+            for (uint256 i = 0; i < specifiedAmounts.length; i++) {
+                uint256 outAmt = ISTAMPL(STAMPL).computeUnderlyingToPerpSwapAmt(
+                    specifiedAmounts[i]
+                );
+                prices[i] = Fraction(outAmt, specifiedAmounts[i]);
+            }
+        } else if (sellToken == SPOT && buyToken == AMPL) {
+            // SPOT -> AMPL
+            for (uint256 i = 0; i < specifiedAmounts.length; i++) {
+                uint256 outAmt = ISTAMPL(STAMPL).computePerpToUnderlyingSwapAmt(
+                    specifiedAmounts[i]
+                );
+                prices[i] = Fraction(outAmt, specifiedAmounts[i]);
+            }
+        } else {
+            revert Unavailable("SwapRouteUndefined");
+        }
+    }
+
+    /**
+     * @dev Computes price quotes for BillBroker swaps.
+     */
+    function _calculatePriceBillBroker(
+        address sellToken,
+        address buyToken,
+        uint256[] memory specifiedAmounts,
+        Fraction[] memory prices
+    ) private {
+        if (sellToken == SPOT && buyToken == USDC) {
+            // SPOT -> USDC
+            for (uint256 i = 0; i < specifiedAmounts.length; i++) {
+                uint256 outAmt = IBillBroker(BILL_BROKER)
+                    .computePerpToUSDSwapAmt(specifiedAmounts[i]);
+                prices[i] = Fraction(outAmt, specifiedAmounts[i]);
+            }
+        } else if (sellToken == USDC && buyToken == SPOT) {
+            // USDC -> SPOT
+            for (uint256 i = 0; i < specifiedAmounts.length; i++) {
+                uint256 outAmt = IBillBroker(BILL_BROKER)
+                    .computeUSDToPerpSwapAmt(specifiedAmounts[i]);
+                prices[i] = Fraction(outAmt, specifiedAmounts[i]);
+            }
+        } else {
+            revert Unavailable("SwapRouteUndefined");
+        }
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                          PRIVATE LIMIT LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Computes swap limits for WAMPL.
+     */
+    function _getWAMPLPoolLimits(
+        uint256[] memory limits,
+        address sellToken,
+        address buyToken
+    ) private view {
+        if (sellToken == AMPL && buyToken == WAMPL) {
+            limits[1] = IERC20(WAMPL).totalSupply() / RESERVE_LIMIT_FACTOR;
```

Comment:
What’s the thinking behind the RESERVE_LIMIT_FACTORs?

Does it guard against anything specific? Or is it simply an indication that something “weird” is going on and we probably shouldn’t allow it.

Looks like it will limit any trade that decreases WAMPL supply more than 90%?

(That answer might also explain why the stampl limit is different also, which I'm curious about)

Reply:
The idea with the factors was just to chose a reasonable under-estimate such that swaps are always guaranteed. (Similar to the univ2 adapter). 

The logic for precisely figuring out the swap limits for stAMPL is complicated. We have liquidity constraints,  drBounds etc. In practice AMPL <> WAMPL doesn't have any limits. And SPOT <> USDC should be liimited by only arHardBounds (though i need to double check).